chalicelib/checks/wfr_checks.py

```diff
@@ -14,22 +14,258 @@
+# _metadata_for_patch
+############################################################
+def _metadata_for_patch(metawfr_uuid, ff_key):
+    """
+    Fetch meta data for metawfr, sample_processing,
+    sample_processing short (frame=raw)
+    """
+    ### Getting metawfr metadata
+    metawfr_meta = ff_utils.get_metadata(metawfr_uuid, key=ff_key)
+    ### Getting the case that will be patched
+    case_acc = metawfr_meta['title'].split(' ')[-1]
+    case_meta = ff_utils.get_metadata(case_acc, add_on='?frame=raw', key=ff_key)
+    ### Getting the sample_processing that will be modified and used for patching
+    sp_uuid = case_meta['sample_processing']
+    sp_meta = ff_utils.get_metadata(sp_uuid, key=ff_key)
+    sp_meta_short = ff_utils.get_metadata(sp_uuid, add_on='?frame=raw', key=ff_key)
+
+    return (metawfr_meta, sp_meta, sp_meta_short)
+
+############################################################
+# _eval_qcs
+############################################################
+def _eval_qcs(metawfr_meta):
+    """
+    Check final_status for metawfr is completed
+    Check expected QCs in overall_qcs are PASS
+    """
+    ### Check final_status
+    if metawfr_meta['final_status'] != 'completed':
+        return False
+
+    ### Check overall_qcs
+    # check is overall_qcs
+    overall_qcs = metawfr_meta.get('overall_qcs', [])
+    if not overall_qcs: # check if overall_qcs exists
+        return False
+
+    # check expected qcs are in overall_qcs
+    expected_qcs = {'linecount_test'}
+    availabl_qcs = set()
+    for qc in overall_qcs:
+        availabl_qcs.add(qc['name'])
+
+    if expected_qcs.difference(availabl_qcs):
+        return False
+
+    # check all expected qcs are pass
+    for qc in overall_qcs:
+        if qc['name'] in expected_qcs and qc['value'] != 'PASS':
+            return False
+
+    return True
+
+#+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+#               Check specific functions
+#+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+############################################################
+# patch_processed_files_to_sample (upstream processed files)
+############################################################
+def patch_processed_files_to_sample(metawfr_uuid, ff_key):
+    """
+    Patch sample_processing with final_bam and sample_gvcf,
+    does not patch completed_processes
+    """
+    ### Getting metadata
+    metawfr_meta, sp_meta, _ = _metadata_for_patch(metawfr_uuid, ff_key)
+    ### Checking QCs are PASS and final_status completed
+    if not _eval_qcs(metawfr_meta): # if check fails, return
+        return
+
+    if len(sp_meta['samples']) == 1: # proband only
+        final_bam, sample_gvcf = '', ''
+        for wfr in metawfr_meta['workflow_runs']:
+            if wfr['name'] == 'workflow_gatk-ApplyBQSR-check':
+                final_bam = wfr['output'][0]['file']['uuid']
+            elif wfr['name'] == 'workflow_gatk-HaplotypeCaller':
```

Comment:
I would make these constants and note what is meaningful about them (their completion indicates pipeline is finished).